### PR TITLE
cleanup handling of hidden dataset schema column "_key"

### PR DESCRIPTION
**Pull request by labkey-matthewb**
#### Rationale
Fixes incompatibilities with new/simpler matchColumns()

#### Related Pull Requests
* 

#### Changes
* Dataset insertRows() returns internal columns "_rowNumber" and "_key".  That seems wrong (and illuminated the real bug)
* We don't ignore "_key" in the input data
* We always add "_key" to the DatasetDIB in order to facilitate generating the lsid column, and filling in the provisioned _key column
* The _key column (returned by insert) and the _key column created by DatasetDIB cause a duplicate column error.

### CODE REVIEW

File: study/src/org/labkey/study/query/DatasetUpdateService.java [hunk 1]
```diff
@@ -283,6 +287,16 @@ protected int _pump(DataIteratorBuilder etl, final ArrayList<Map<String, Object>
         {
             boolean hasRowId = _dataset.getKeyManagementType() == Dataset.KeyManagementType.RowId;
 
+            if (null != rows)
+            {
+                // TODO: consider creating DataIterator metadata to mark "internal" cols (not to be returned via API)
```
Comment (labkey-matthewb) [inline study/src/org/labkey/study/query/DatasetUpdateService.java:292]: This change is not strictly necessary.  I was wondering how to get rid of the _rowNumber value that is returned by the insertRows() API.

File: api/src/org/labkey/api/dataiterator/DataIteratorUtil.java [hunk 2]
```diff
@@ -227,39 +223,15 @@ public static boolean isEtl(@NotNull DataIteratorContext context)
...
-                to = targetMap.get(from.getName());
-                if (null == to && null != from.getPropertyURI())
-                {
-                    // Do we actually rely on this anywhere???
-                    // Like maybe ETL from one study to another where subject name does not match? or assay publish?
-                    to = targetMap.get(from.getPropertyURI());
-                    if (null != to)
-                        to = new Pair<>(to.first, org.labkey.api.dataiterator.DataIteratorUtil.MatchType.low);
-                }
+                // Do we actually rely on this anywhere???
```
Comment (labkey-matthewb) [inline api/src/org/labkey/api/dataiterator/DataIteratorUtil.java:229]: BTW, I'm running tests to try to answer this question

**Comment by labkey-martyp** [comment]
I kicked off some tests on TC on this branch.  Looks like some of the study tests are failing the dataset key assertion you added.
[StudyDateBasedTest](https://teamcity.labkey.org/viewLog.html?buildId=1416235&tab=buildResultsDiv&buildTypeId=LabkeyTrunk_DailySuites_DailyESqlserver&branch_LabkeyTrunk_DailySuites=key_cleanup)
[StudyContinuousTest](https://teamcity.labkey.org/viewLog.html?buildId=1416234&tab=buildResultsDiv&buildTypeId=LabkeyTrunk_DailySuites_DailyDSqlserver&branch_LabkeyTrunk_DailySuites=key_cleanup)

**Review by labkey-matthewb — COMMENTED** [review]
> I kicked off some tests on TC on this branch. Looks like some of the study tests are failing the dataset key assertion you added.
> [StudyDateBasedTest](https://teamcity.labkey.org/viewLog.html?buildId=1416235&tab=buildResultsDiv&buildTypeId=LabkeyTrunk_DailySuites_DailyESqlserver&branch_LabkeyTrunk_DailySuites=key_cleanup)
> [StudyContinuousTest](https://teamcity.labkey.org/viewLog.html?buildId=1416234&tab=buildResultsDiv&buildTypeId=LabkeyTrunk_DailySuites_DailyDSqlserver&branch_LabkeyTrunk_DailySuites=key_cleanup)

thanks fixed assert

File: api/src/org/labkey/api/dataiterator/DataIteratorUtil.java (hunk 2 above)
Comment (labkey-martyp) [reply api/src/org/labkey/api/dataiterator/DataIteratorUtil.java:229]: This [comment ](https://github.com/LabKey/platform/blob/696a7d7af9279272004e03310feec9949af7a29c/study/src/org/labkey/study/model/DatasetDefinition.java#L2045) makes me wonder about study import/publish.  Although the data TSVs should only have column names.

File: study/src/org/labkey/study/query/DatasetUpdateService.java (hunk 1 above)
Comment (labkey-martyp) [reply study/src/org/labkey/study/query/DatasetUpdateService.java:292]: I think you mentioned we could also add something to ColumnInfos to mark internal columns.

**Review by labkey-martyp — APPROVED** [review]

**Pull request merged**